metadata/src/main/java/org/apache/kafka/image/publisher/BrokerRegistrationTracker.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.image.publisher;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.image.MetadataDelta;
+import org.apache.kafka.image.MetadataImage;
+import org.apache.kafka.image.loader.LoaderManifest;
+import org.apache.kafka.metadata.BrokerRegistration;
+import org.apache.kafka.server.common.MetadataVersion;
+import org.slf4j.Logger;
+
+import java.util.List;
+
+/**
+ * Tracks the registration of a specific broker, and executes a callback if it should be refreshed.
+ *
+ * This tracker handles cases where we might want to re-register the broker. The only such case
+ * right now is during the transition from non-JBOD mode, to JBOD mode. In other words, the
+ * transition from a MetadataVersion less than 3.7-IV2, to one greater than or equal to 3.7-IV2.
+ * In this case, the broker registration will start out containing no directories, and we need to
+ * resend the BrokerRegistrationRequest to fix that.
+ *
+ * As much as possible, the goal here is to keep things simple. We just compare the desired state
+ * with the actual state, and try to make changes only if necessary.
+ */
+public class BrokerRegistrationTracker implements MetadataPublisher {
```

Comment:
I'm wondering if there is a more general case here that would be useful to extract – listening for metadata changes to enable features. If storage has been formatted a long time ago, the MV at statup can be old and the broker will only find out about the current MV as it catches up with metadata. JBOD won't be the only feature that needs to be activated upon discovering a newer MV. Maybe there aren't many of these cases now, but it would better if all those features get activated via the same metadata listening mechanism. What do you think?

Reply:
Yeah. I like the idea and I think we might want to add code like that here later.